[B+C] Added middle click detection for InventoryClickEvent. Fixes BUKKIT-3692

THDigi opened the pull request:
Adds a way to detect middle clicks in InventoryClickEvent.

CB PR: https://github.com/Bukkit/CraftBukkit/pull/1049
Ticket: https://bukkit.atlassian.net/browse/BUKKIT-3692

riking commented on src/main/java/org/bukkit/event/inventory/InventoryClickEvent.java:
```diff
@@ -58,12 +64,19 @@ public ItemStack getCurrentItem() {
         return rightClick;
     }
-
+    
```
4 spaces. Delete.

riking commented on src/main/java/org/bukkit/event/inventory/InventoryClickEvent.java:
```diff
@@ -58,12 +64,19 @@ public ItemStack getCurrentItem() {
+    public boolean isMiddleClick() {
+        return middleClick;
+    }
+    
```
Spaces. Delete.

desht commented:
Now that there are potentially three buttons to deal with, I wonder if this would be better handled with an enum, e.g. enum MouseButton { LEFT, MIDDLE, RIGHT};

In particular, right now it's possible to create this event with both 'middle' and 'right' set to true, which is not a valid state (the protocol does not allow for more than one button to be pressed simultaneously).

THDigi commented:
@desht I initially wanted to just parse the button value to the event then check it there, but I guess an enum would be even better, I'll do that.
And yes, that's true, another reason to use enum.
EDIT: I'll make the enum as an inner class, tell me if I must make it a diferent class and where should I put it.

@riking I still can't fully understand git, can you tell me how I can do that ?
My last attempt merged codes and I had duplicated isMiddleClick()...

desht commented:
@THDigi this is a good reference for managing github PR's: https://openshift.redhat.com/community/wiki/github-workflow-for-submitting-pull-requests

THDigi commented:
After 2 hours of struggling with squashing I found I can use something else: reset, new commit, force push.

desht commented:
That works too!  Really hope this gets accepted, I already have some ideas on how to use it :)

SuperSpyTX commented:
This would introduce some interesting opportunities for plugin developers.

turt2live commented:
If it were me I'd deprecate is(Right/Left)Click() because of getButton().

THDigi commented:
@desht @SuperSpyTX Indeed!

@turt2live Oh leave them be, they're just shortcuts, most events have them.

By the way, is a PR required to remove old deprecated code or is it removed after a certain amount of time by the staff ?

EdGruberman commented:
Deprecated code would be removed at a later date by anyone as appropriate.  Deprecating code simply points newer developers to use the preferred code and notifies existing developers of a possible future breaking change.

jrtc27 commented on src/main/java/org/bukkit/event/inventory/InventoryClickEvent.java:
```diff
@@ -11,16 +11,26 @@
     private ItemStack current = null;
 
+    public enum MouseButton {
+        LEFT, MIDDLE, RIGHT;
```
Should be on separate lines to adhere to Bukkit's formatting standards.

THDigi replied:
Pff, this could've been avoided if a fixed automatic standard were to be released (IDE formatter file).

Wolvereness commented:
If the concept of more than L/R click type exist, then the event needs to be future-aware. Plugins should anticipate the possibility of a fourth, fifth, etc. click type, and this PR has no way to address that. Having a method for each 'type' to return a boolean isn't a good idea if you consider fourth and fifth clicks that are not implemented.

There seems to be a real lack of awareness of what's going on (as far as the event is concerned). Can this be improved at all? Although I have no used the API myself, this really looks like it relies on pre-defined client & server behavior, but there's no indication this is provided. In light of this and personal opinion, the proposed change appears to be a bit of a placebo.

desht commented:
@Wolvereness I'm not the originator of the PR, but I think it's more than just a placebo.  Middle-clicking is something that's in the Minecraft protocol now, and it would be extremely useful for plugins to be able to detect it.

Having said that, future-proofing is a very good point.  But what conventions are you going to use to number the buttons?  Following the Minecraft client terminology makes most sense (the only sense IMHO), so you'd have MouseButton.BUTTON1 for left-click, MouseButton.BUTTON2 for right-click, and MouseButton.BUTTON3 for middle-click.  If future versions of the protocol add other mouse buttons, the enum can be extended.

I agree that the separate isLeftClick(), isRightClick() methods should be deprecated in that case, and no new isMiddleClick() method should be added.  Instead plugins should move to using event.getMouseButton().

THDigi commented:
@Wolvereness 
I'm trying to contribute, if you want it designed diferently then please tell and I'll change it, but you must understand that the current event's click detection is a bit broken when it comes to right and middle clicks because it can't tell them apart.

The protocol currently specifies that its button values are 0 = left, 1 = right and 3 = middle, my initial thought was to just get the button number and pass it through the event, that would surely have forward compatibility without any changes to the API (unless the protocol gets its numbers swapped around) but that won't be as nice as enums, which can be easly updated by adding an enum to this and a case to the switch() from the CB code.

The isLeft/RightClick() methods were already there, I just added isMiddleClick() to be consistent with the rest of the code, but I guess they will be a bit weird once it gets to isSideClick() or stuff like that... so I'll make them deprecated and remove my isMiddleClick().

THDigi committed:
Added middle click detection for InventoryClickEvent [BUKKIT-3692]

riking commented on src/main/java/org/bukkit/event/inventory/InventoryClickEvent.java:
```diff
@@ -11,16 +11,40 @@
+         */
+        LEFT,
+        /**
+         * Middle mouse button or wheel click
```
Maybe document that scrollwheel does not apply for this event:

```
Middle mouse button or wheel click
<p />
Note that the scrollwheel is represented in a different event, {@link org.bukkit.event.PlayerItemHeldEvent}
```

THDigi replied:
For character actions they are all represented by other events but not inventory events, should I make references to PlayerItemHeldEvent for middle and PlayerInteractEvent for left/right ?

steveanton replied:
The enum strictly refers to mouse **buttons**, so it would be reasonable to assume that the reader would know that the scroll **wheel** does not apply.

riking replied:
@THDigi Oh yeah, that's right. Hm, yeah, don't change ti.

THDigi committed:
Changed javadoc of isShiftClick() to inform of Shift+MiddleClick behaviour

When client shift+middleclicks the packet he sends will not contain the
shift bit, the server can't fix for this so we must inform the API users
of this behaviour.

riking commented:
Rejected as an incomplete fix to a much larger problem
